### ISSUE TITLE
Expose cached versions of resources during events.

### ISSUE THREAD

**Issue by macdja38** [qeled/discordie#44]
During events like: http://qeled.github.io/discordie/#/docs/Events it would be great if we could get for example the previous version of the role exposed by GUILD_ROLE_UPDATE

key | contents
--- | ---
socket | GatewaySocket 
guild | IGuild
role | IRole
previousRole | IRole \| null

It would allow us to create features like a moderation log that logs changes to assets instead of just the new version. Without having to maintain an expensive and troubling cach of **everything**.

Events I would like to see this added to if possible:
- [x] CHANNEL_UPDATE
- [x] GUILD_EMOJIS_UPDATE <-- not sure why this is a seporate event?
- [x] GUILD_UPDATE
- [x] GUILD_MEMBER_UPDATE <-- has a partial implementation
- [x] GUILD_ROLE_UPDATE

**Comment by qeled** [comment]
[First implementation](https://github.com/qeled/discordie/commit/5e1b447449c4277236dd52ce2b89b63cfdf5b77b) pushed to the [update-events-changes](https://github.com/qeled/discordie/tree/update-events-changes) branch.

Providing a "previous interface" is not quite possible without rewriting the base because it always has to point to the latest object.

Currently it exposes a property with a function that converts internal model into raw object in format that Discord provides:

```js
client.Dispatcher.on("CHANNEL_UPDATE", e => {
  const data = e.getChanges();
  console.log(data.before);
  console.log(data.after);
  
  // before:
  //  { id: '...',
  //    name: 'abc',
  //    topic: '',
  //    permission_overwrites: [
  //      { id: '...', type: 'role', allow: 0, deny: 0 }
  //    ],
  //    ... }
  // after:
  //  { id: '...',
  //    name: 'abc',
  //    topic: 'new topic',
  //    permission_overwrites: [
  //      { id: '...', type: 'role', allow: 0, deny: 0 }
  //    ],
  //    ... }
})
```

Further naming/design suggestions are welcome.

**Comment by Meowbeast** [comment]
This is great!  Thanks @qeled

**Comment by macdja38** [comment]
Looks absolutely Perfect, I'll try this out over the next few days to week and report back.

**Comment by briantanner** [comment]
Hey @qeled has this been implemented or is there an update? I'm starting to use more events than I have in the past now, and one specifically, `ROLE_DELETE`, doesn't give any reference to the deleted role, only a role id according to the docs.

Also, I could really use access to cached objects on other DELETE events as well, specifically `GUILD_DELETE`, where I'm looking for member_count, among other cached data. That's not given in the raw data from discord.

Thanks

**Comment by qeled** [comment]
Released in 0.9.0.
